src/appservice/Appservice.ts

```diff
@@ -658,6 +668,70 @@ export class Appservice extends EventEmitter {
         return providedToken === this.registration.hs_token;
     }
 
+    private async decryptAppserivceEvent(roomId: string, encrypted: EncryptedRoomEvent): ReturnType<Appservice["processEvent"]> {
+        const existingClient = this.cryptoClientForRoomId.get(roomId);
+        const decryptFn = async (client: MatrixClient) => {
+            let event = (await client.crypto.decryptRoomEvent(encrypted, roomId)).raw;
```

Comment:
If the rust-sdk's `OlmMachine` doesn't discard keys for rooms that a user has left, then this should first check if `client` is currently in the target room.  Otherwise, this would decrypt an event that `client` shouldn't be able to access.

On the other hand, it might not matter in practice whether `client` shouldn't see the event, because as long as _any_ of the appservice's users can see it, there's no difference in terms of privacy/security which of its clients the appservices uses to decrypt an event.

Reply:
I don't think we materially care on the AS side. If we still have keys and can turn an encrypted body into unencrypted text then yay. It doesn't really matter which client did it.